isis/tests/FunctionalTestsMsi2isis.cpp

```diff
@@ -0,0 +1,469 @@
+  QVector<QString> args = {"from=data/near/msi2isis/m0157063318f4_2p_iof_gbl.lbl",
+                            "to=" + cubeFileName };
+
+  UserInterface options(APP_XML, args);
+  try {
+    std::cerr << "Running msi2isis..." << std::endl;
+    msi2isis(options, &appLog);
+    std::cerr << "Success!" << std::endl;
+
+  }
+  catch (IException &e) {
+    FAIL() << "Unable to ingest NEAR/MSI GBL file: " << e.toString().toStdString().c_str() << std::endl;
+  }
+  Cube cube(cubeFileName);
+  Pvl *isisLabel = cube.label();
+
+  std::cerr << "Got labels... "<< std::endl;
+
+  // Dimensions Group
+  ASSERT_EQ(cube.sampleCount(), 537);
+  ASSERT_EQ(cube.lineCount(), 412);
+  ASSERT_EQ(cube.bandCount(), 1);
+
+  // Instrument Group
+  PvlGroup &inst = isisLabel->findGroup("Instrument", Pvl::Traverse);
+  ASSERT_EQ(inst["SpacecraftName"][0].toStdString(), "NEAR EARTH ASTEROID RENDEZVOUS");
+  ASSERT_EQ(inst["InstrumentId"][0].toStdString(), "MSI" );
+  ASSERT_EQ(inst["TargetName"][0].toStdString(), "EROS" );
+  ASSERT_EQ(inst["SpacecraftClockStartCount"][0].toStdString(), "157063318856" );
+  ASSERT_EQ(inst["SpacecraftClockStopCount"][0].toStdString(), "157063318919" );
+  ASSERT_EQ(inst["OriginalSpacecraftClockStartCount"][0].toStdString(), "157063318.856" );
+  ASSERT_EQ(inst["OriginalSpacecraftClockStopCount"][0].toStdString(), "157063318.919" );
+  ASSERT_DOUBLE_EQ(double(inst["ExposureDuration"]), 63.0);
+  ASSERT_EQ(inst["StartTime"][0].toStdString(), "2001-02-08T17:24:42.217" );
+  ASSERT_EQ(inst["StopTime"][0].toStdString(), "2001-02-08T17:24:42.280" );
+  ASSERT_DOUBLE_EQ(double(inst["DpuDeckTemperature"]), 286.5);
+  ASSERT_EQ(inst["DpuDeckTemperature"].unit().toStdString(), "K");
+
+  // Archive Group
+  PvlGroup &archive = isisLabel->findGroup("Archive", Pvl::Traverse);
+  ASSERT_EQ(archive["DataSetId"][0].toStdString(), "NEAR-A-MSI-3-EDR-EROS/ORBIT-V1.0");
+  ASSERT_EQ(archive["ProductId"][0].toStdString(), "M0157063318F4_2P_IOF_GBL.FIT");
+  ASSERT_EQ(archive["InstrumentHostName"][0].toStdString(), "NEAR EARTH ASTEROID RENDEZVOUS");
+  ASSERT_EQ(archive["InstrumentName"][0].toStdString(), "MULTI-SPECTRAL IMAGER");
+  ASSERT_EQ(archive["InstrumentId"][0].toStdString(), "MSI");
+  ASSERT_EQ(archive["TargetName"][0].toStdString(), "EROS");
+
+  // BandBin Group
+  PvlGroup &bandbin = isisLabel->findGroup("BandBin", Pvl::Traverse);
+  ASSERT_EQ(int(bandbin["FilterNumber"]), 4);
+  ASSERT_EQ(int(bandbin["Center"]), 950);
+  ASSERT_EQ(bandbin["Center"].unit().toStdString(), "nm");
+
+ // Kernels Group
+  PvlGroup &kernel = isisLabel->findGroup("Kernels", Pvl::Traverse);
+  ASSERT_EQ(int(kernel["NaifFrameCode"]), -93001);
+
+  std::unique_ptr<Histogram> hist (cube.histogram());
+
+  ASSERT_NEAR(hist->Average(), 0.042633522648316, .000001);
+  ASSERT_NEAR(hist->Sum(), 6947.8146519054, .0001);
+  ASSERT_EQ(hist->ValidPixels(), 162966);
+  ASSERT_EQ(hist->NullPixels(), 58278);
+  ASSERT_NEAR(hist->StandardDeviation(), 0.016649588425538, .0001);
```

Comment:
Most of these ASSERTs can be changed to EXPECTs. Generally, I only use a hard ASSERT when it would result in the following test code to fail. The most common examples of this are testing the size of output before iterating over it and checking that a PVL group/object exists before checking its contents.

Reply:
Done.